File: action_ws/src/mcp2515_can_driver/src/mcp2515_can_client.cpp
#include <ros/ros.h>
#include <actionlib/client/simple_action_client.h>
#include <mcp2515_can_driver/MCP2515CANCommAction.h>
#include <iostream>
#include <iomanip>
#include <vector>

void feedbackCallback(const mcp2515_can_driver::MCP2515CANCommFeedbackConstPtr& feedback)
{
    ROS_INFO("Feedback: %s", feedback->status.c_str());
}

int main(int argc, char** argv)
{
    ros::init(argc, argv, "mcp2515_can_client");
    ROS_INFO("MCP2515 CAN Client started");

    std::string action_name = "mcp2515_can_comm";
    actionlib::SimpleActionClient<mcp2515_can_driver::MCP2515CANCommAction> ac(action_name, true);

    ROS_INFO("Waiting for action server...");
    if (!ac.waitForServer(ros::Duration(10.0))) {
        ROS_ERROR("ERROR: Action server not available");
        return -1;
    }
    ROS_INFO("OK: Connected to action server");

    int can_id = 0x123;
    int dlc = 4;
    std::vector<uint8_t> data_vec = {0x11, 0x22, 0x33, 0x44, 0x00, 0x00, 0x00, 0x00};
    bool extended = false;
    bool remote = false;

    ros::NodeHandle nh("~");
    nh.param("can_id", can_id, 0x123);
    nh.param("dlc", dlc, 4);
    nh.param("extended", extended, false);
    nh.param("remote", remote, false);

    std::string id_type = extended ? "EXT" : "STD";
    std::string frame_type = remote ? "RTR" : "DATA";
    ROS_INFO("Sending goal: %s ID=0x%X %s DLC=%u", id_type.c_str(), can_id,
             frame_type.c_str(), dlc);

    mcp2515_can_driver::MCP2515CANCommGoal goal;
    goal.can_id = can_id;
    goal.dlc = dlc;
    for (size_t i = 0; i < data_vec.size() && i < 8; ++i) {
        goal.data[i] = data_vec[i];
    }
    goal.extended = extended;
    goal.remote = remote;

    ac.sendGoal(goal, NULL, NULL, boost::bind(&feedbackCallback, _1));

    ROS_INFO("Waiting for result...");
    bool finished = ac.waitForResult(ros::Duration(10.0));

    if (finished) {
        actionlib::SimpleClientGoalState state = ac.getState();
        ROS_INFO("Action finished: %s", state.toString().c_str());

        mcp2515_can_driver::MCP2515CANCommResultConstPtr result = ac.getResult();
        if (result) {
            ROS_INFO("Result:");
            ROS_INFO("   success: %s", result->success ? "true" : "false");
            ROS_INFO("   message: %s", result->message.c_str());

            if (result->success) {
                std::cout << "   received_data: [";
                for (size_t i = 0; i < result->received_data.size(); ++i) {
                    std::cout << "0x" << std::hex << std::setw(2) << std::setfill('0')
                              << (int)result->received_data[i] << std::dec;
                    if (i < result->received_data.size() - 1) std::cout << ", ";
                }
                std::cout << "]" << std::endl;
                ROS_INFO("   received_id: 0x%X", result->received_id);
                ROS_INFO("   received_dlc: %u", result->received_dlc);
            }
        }
    } else {
        ROS_WARN("Action did not finish before timeout");
    }

    ROS_INFO("Client test complete");
    return 0;
}

File: action_ws/src/mcp2515_can_driver/scripts/mcp2515_can_client.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
=======================================
MP2515 SPI转CAN 客户端
=======================================
功能说明:
  - 向 MP2515 CAN Action Server 发送 CAN 帧
  - 支持标准帧和扩展帧
  - 支持数据帧和远程帧
  
使用方法:
  source devel/setup.bash
  rosrun mcp2515_can_driver mp2515_can_client.py
  
作者: Jetson Nano
日期: 2026-04-08
"""

import rospy
import actionlib
from mcp2515_can_driver.msg import MCP2515CANCommAction, MCP2515CANCommGoal, MCP2515CANCommResult


class MP2515CANClient:
    """MP2515 CAN 通信客户端"""
    
    def __init__(self):
        self.server_name = 'mp2515_can_comm'
        rospy.loginfo(f"正在连接到 {self.server_name}...")
        
        self.client = actionlib.SimpleActionClient(self.server_name, MCP2515CANCommAction)
        self.client.wait_for_server()
        
        rospy.loginfo(f"✅ 已连接到 {self.server_name}")
    
    def send_can_frame(self, can_id, data, dlc, extended=False, remote=False, timeout=5.0):
        """
        发送 CAN 帧
        
        参数:
          can_id: CAN ID (11-bit 标准 或 29-bit 扩展)
          data: CAN 数据 (列表)
          dlc: 数据长度 (0-8)
          extended: 是否为扩展帧
          remote: 是否为远程帧
          timeout: 超时时间(秒)
        
        返回:
          success: 是否成功
          message: 状态消息
          result: 完整结果对象
        """
        goal = MCP2515CANCommGoal()
        goal.can_id = can_id
        goal.dlc = dlc
        goal.data = data
        goal.extended = extended
        goal.remote = remote
        
        id_type = "EXT" if extended else "STD"
        frame_type = "RTR" if remote else "DATA"
        rospy.loginfo(f"📤 发送 CAN 帧: {id_type} ID=0x{can_id:X} {frame_type} DLC={dlc}")
        
        self.client.send_goal(goal)
        
        self.client.wait_for_result(rospy.Duration(timeout))
        result = self.client.get_result()
        
        if result:
            rospy.loginfo(f"✅ 结果: success={result.success}, message={result.message}")
        else:
            rospy.logerr("❌ 未收到结果 (超时)")
        
        return result.success if result else False, result.message if result else "超时", result
    
    def feedback_cb(self, feedback):
        """反馈回调"""
        rospy.loginfo(f"📋 反馈: {feedback.status}")


def main():
    rospy.init_node('mp2515_can_client')
    
    client = MP2515CANClient()
    
    rospy.loginfo("发送测试 CAN 帧...")
    
    success, msg, result = client.send_can_frame(
        can_id=0x123,
        data=[0x11, 0x22, 0x33, 0x44, 0, 0, 0, 0],
        dlc=4,
        extended=False,
        remote=False
    )
    
    if success:
        rospy.loginfo("✅ 测试帧发送成功")
    else:
        rospy.logerr(f"❌ 测试帧发送失败: {msg}")
    
    rospy.loginfo("客户端测试完成")


if __name__ == '__main__':
    try:
        main()
    except rospy.ROSInterruptException:
        rospy.loginfo("📛 收到中断信号")
    except Exception as e:
        rospy.logerr(f"❌ 客户端错误: {e}")
